File: src/enable_robot_by_button.cpp
//################################################################
//# Copyright (C) 2024, EduArt Robotik GmbH, All rights reserved.#
//# Further information can be found in the LICENSE file.        #
//################################################################

#include <rclcpp/rclcpp.hpp>

#include <edu_robot/srv/set_mode.hpp>

#include <mraa/gpio.hpp>

using namespace std::chrono_literals;

// helper function for getting given mode as string
std::string get_mode_string(const edu_robot::msg::Mode mode)
{
  std::string mode_string;

  if (mode.mode & edu_robot::msg::Mode::INACTIVE) {
    mode_string += "INACTIVE|";
  }
  if (mode.mode & edu_robot::msg::Mode::REMOTE_CONTROLLED) {
    mode_string += "REMOTE CONTROLLED|";
  }
  if (mode.mode & edu_robot::msg::Mode::AUTONOMOUS) {
    mode_string += "FLEET|";
  }
  if (mode.feature_mode & edu_robot::msg::Mode::COLLISION_AVOIDANCE) {
    mode_string += "COLLISION_AVOIDANCE|";
  }
  if (mode.feature_mode & edu_robot::msg::Mode::COLLISION_AVOIDANCE_OVERRIDE) {
    mode_string += "COLLISION_AVOIDANCE_OVERRIDE|";
  }  
  if (mode.drive_kinematic & edu_robot::msg::Mode::SKID_DRIVE) {
    mode_string += "SKID_DRIVE|";
  }
  if (mode.drive_kinematic & edu_robot::msg::Mode::MECANUM_DRIVE) {
    mode_string += "MECANUM_DRIVE|";
  }

  if (mode_string.empty() == false) {
    mode_string.pop_back();
  }

  return mode_string;
}

/**
 * \brief This ROS node enables the robot by a button pressure at a GPIO. It shows how a button can be read as input.
 *
 * This node reads a button as input and triggers an ROS2 service call. The GPIO is controlled by the library MRAA.
 */
class EnableRobotByButton : public rclcpp::Node
{
public:
  EnableRobotByButton()
    : rclcpp::Node("enable_robot_by_button")
    , _enable_button(4) // GPIO PIN 4
  {
    // Create service client for setting mode.
    _client_set_mode = create_client<edu_robot::srv::SetMode>("set_mode");

    // Initialize GPIO pin as input.
    if (mraa::SUCCESS != _enable_button.dir(mraa::DIR_IN)) {
      throw std::runtime_error("Can't set direction for button input.");
    }

    // Start timer: the timer calls the bound callback method/function with given time period of 100ms.
    _timer = create_timer(100ms, std::bind(&EnableRobotByButton::process, this));
  }

private:
  void process()
  {
    // alias for service response future type (just for better reading)
    using ResponseFuture = rclcpp::Client<edu_robot::srv::SetMode>::SharedFutureWithRequest;

    // is called by the timer every 100ms
    const int button_value = _enable_button.read();

    if (button_value < 0) {
      // error occurred during reading of button input value
      RCLCPP_ERROR(get_logger(), "error occurred during reading of button input value");
      return;
    }

    // reading button state: when button is pressed it pulls the signal down to zero voltage.
    const bool is_pressed = button_value == 0;

    if (_was_pressed == false && is_pressed == true) {
      // detected positive edge --> button was pressed right now
      RCLCPP_INFO(get_logger(), "button was pressed");

      auto request = std::make_shared<edu_robot::srv::SetMode::Request>();
      // set robot mode to autonomous
      request->mode.mode = edu_robot::msg::Mode::AUTONOMOUS;

      // call service async and process response in a lambda function
      _client_set_mode->async_send_request(request, [logger = get_logger()](ResponseFuture future){
        const auto response = future.get();

        if ((response.second->state.mode.mode & response.first->mode.mode) == false) {
          RCLCPP_ERROR_STREAM(logger, "Can't disable robot! Robot is in mode = " << get_mode_string(response.second->state.mode));
          return;
        }

        RCLCPP_INFO(logger, "Set mode AUTONOMOUS successfully.");
        RCLCPP_INFO(logger, "Current mode of the robot is = %s", get_mode_string(response.second->state.mode).c_str());      
      });
    }

    _was_pressed = is_pressed;
  }

  std::shared_ptr<rclcpp::Client<edu_robot::srv::SetMode>> _client_set_mode;
  std::shared_ptr<rclcpp::TimerBase> _timer;
  mraa::Gpio _enable_button;
  bool _was_pressed = false;
};

int main(int argc, char** argv)
{
  // Initializing ROS
  rclcpp::init(argc, argv);
  // Instantiating
  rclcpp::spin(std::make_shared<EnableRobotByButton>());
  // Shuting down ROS after the ROS node exited
  rclcpp::shutdown();
}

File: scripts/enable_robot_by_button.py
#!/usr/bin/env python3

import rclpy
from rclpy.node import Node

from edu_robot.srv import SetMode
from edu_robot.msg import Mode
import mraa

# helper function for getting given mode as string
def get_mode_string(mode : Mode) -> str:
  mode_string = str()

  if mode.mode & Mode.INACTIVE:
    mode_string += 'INACTIVE'
  if mode.mode & Mode.REMOTE_CONTROLLED:
    mode_string += 'REMOTE CONTROLLED|'
  if mode.mode & Mode.AUTONOMOUS:
    mode_string += 'FLEET|'
  if mode.mode & Mode.COLLISION_AVOIDANCE:
    mode_string += 'COLLISION_AVOIDANCE|'
  if mode.mode & Mode.COLLISION_AVOIDANCE_OVERRIDE:
    mode_string += 'COLLISION_AVOIDANCE_OVERRIDE|'
  if mode.mode & Mode.SKID_DRIVE:
    mode_string += 'SKID_DRIVE|'
  if mode.mode & Mode.MECANUM_DRIVE:
    mode_string += 'MECANUM_DRIVE|'
  
  if len(mode_string) > 0:
    mode_string = mode_string[:-1]

  return mode_string


"""This ROS node enables the robot by a button pressure at a GPIO. It shows how a button can be read as input.
This node reads a button as input and triggers an ROS2 service call. The GPIO is controlled by the library MRAA.
"""
class EnableRobotByButton(Node):
  def __init__(self):
    # Calling Constructor of Base Class: Important!
    super().__init__('enable_robot_by_button')

    # Create service client for setting mode.
    self.client_set_mode = self.create_client(SetMode, 'set_mode')

    # GPIO 4
    self.was_pressed = False
    self.enable_button = mraa.Gpio(4)
    # Initialize GPIO pin as input.
    self.enable_button.dir(mraa.DIR_IN)

    # Start timer: the timer calls the bound callback method/function with given time period of 100ms.
    self.timer = self.create_timer(0.1, self.process)

  def process(self):
    button_value = self.enable_button.read()

    if button_value < 0:
      # error occurred during reading of button input value
      self.get_logger().error('error occurred during reading of button input value')
      return
    
    # reading button state: when button is pressed it pulls the signal down to zero voltage.
    is_pressed = button_value == 0

    if self.was_pressed == False and is_pressed == True:
      # detected positive edge --> button was pressed right now
      self.get_logger().info('button was pressed"')

      request = SetMode.Request()
      request.mode.mode = Mode.AUTONOMOUS

      self.client_set_mode.call_async(request)

    self.was_pressed = is_pressed


def main():
  rclpy.init()

  enable_robot_by_button_node = EnableRobotByButton()
  rclpy.spin(enable_robot_by_button_node)

  # Destroy the node explicitly
  # (optional - otherwise it will be done automatically when the garbage collector destroys the node object)
  enable_robot_by_button_node.destroy_node()
  rclpy.shutdown()

if __name__ == '__main__':
    main()
